runtime/storage/storage.go

```diff
@@ -0,0 +1,174 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/rilldata/rill/runtime/pkg/gcputil"
+	"gocloud.dev/blob"
+	"gocloud.dev/blob/gcsblob"
+	"gocloud.dev/gcp"
+)
+
+type Client struct {
+	dataDirPath  string
+	bucketConfig *gcsBucketConfig
+	prefixes     []string
+}
+
+func New(dataDir string, bucketCfg map[string]any) (*Client, error) {
+	c := &Client{
+		dataDirPath: dataDir,
+	}
+
+	if len(bucketCfg) != 0 {
+		gcsBucketConfig := &gcsBucketConfig{}
+		err := mapstructure.WeakDecode(bucketCfg, gcsBucketConfig)
+		if err != nil {
+			return nil, err
+		}
+		c.bucketConfig = gcsBucketConfig
+	}
+	return c, nil
+}
+
+func MustNew(dataDir string, bucketCfg map[string]any) *Client {
+	c, err := New(dataDir, bucketCfg)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func (c *Client) WithPrefix(prefix ...string) *Client {
+	newClient := &Client{
+		dataDirPath:  c.dataDirPath,
+		bucketConfig: c.bucketConfig,
+	}
+	newClient.prefixes = append(newClient.prefixes, c.prefixes...)
+	newClient.prefixes = append(newClient.prefixes, prefix...)
+	return newClient
+}
+
+func (c *Client) RemovePrefix(ctx context.Context, prefix ...string) error {
+	if c.prefixes != nil {
+		return fmt.Errorf("storage: RemovePrefix is not supported for prefixed client")
+	}
+
+	// clean data dir
+	removeErr := os.RemoveAll(c.path(c.dataDirPath, prefix...))
+
+	// clean temp dir
+	removeErr = errors.Join(removeErr, os.RemoveAll(c.path(os.TempDir(), prefix...)))
+
+	// clean bucket
+	bkt, ok, err := c.OpenBucket(ctx, prefix...)
+	if err != nil {
+		return errors.Join(removeErr, err)
+	}
+	if !ok {
+		return removeErr
+	}
+	defer bkt.Close()
+
+	iter := bkt.List(&blob.ListOptions{})
+	for {
+		obj, err := iter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return errors.Join(removeErr, err)
+		}
+		err = bkt.Delete(ctx, obj.Key)
+		if err != nil {
+			return errors.Join(removeErr, err)
+		}
+	}
+	return removeErr
+}
+
+func (c *Client) DataDir(elem ...string) (string, error) {
+	path := c.path(c.dataDirPath, elem...)
+	err := os.MkdirAll(path, os.ModePerm)
+	if err != nil {
+		return "", err
+	}
+	return path, nil
+}
+
+func (c *Client) TempDir(elem ...string) (string, error) {
+	path := c.path(os.TempDir(), elem...)
```

Comment:
I don't believe `os.TempDir` is guaranteed to be unique for this process. So it probably needs to call something like `os.MkdirTemp("", "rill")` in `New` and store that directory in the `Client`. 

This is to avoid collision with other temp files, and to ensure that listing `TempDir` on the root `storage.Client` can be used to capture metrics about temp data usage for all instances.